File: src/main/java/org/moinex/ui/dialog/investment/BaseTickerManagement.java
/*
 * Filename: BaseTickerManagement.java
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.AssetType;
import org.moinex.service.PreferencesService;
import org.moinex.service.TickerService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.springframework.beans.factory.annotation.Autowired;

/** Base class to implement common methods for AddTickerController and EditTickerController */
@NoArgsConstructor
public abstract class BaseTickerManagement {
    @FXML protected TextField nameField;

    @FXML protected TextField symbolField;

    @FXML protected TextField currentPriceField;

    @FXML protected TextField quantityField;

    @FXML protected TextField avgUnitPriceField;

    @FXML protected ComboBox<AssetType> typeComboBox;

    protected TickerService tickerService;
    protected PreferencesService preferencesService;

    /**
     * Constructor
     *
     * @param tickerService Ticker service
     * @note This constructor is used for dependency injection
     */
    @Autowired
    protected BaseTickerManagement(
            TickerService tickerService, PreferencesService preferencesService) {
        this.tickerService = tickerService;
        this.preferencesService = preferencesService;
    }

    @FXML
    protected void initialize() {
        configureComboBoxes();
        populateTypeComboBox();
        configureListeners();
    }

    @FXML
    protected void handleCancel() {
        Stage stage = (Stage) nameField.getScene().getWindow();
        stage.close();
    }

    @FXML
    protected abstract void handleSave();

    protected void populateTypeComboBox() {
        typeComboBox.getItems().setAll(AssetType.values());
    }

    protected void configureComboBoxes() {
        UIUtils.configureComboBox(
                typeComboBox, t -> UIUtils.translateAssetType(t, preferencesService));
    }

    protected void configureListeners() {
        currentPriceField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.INVESTMENT_VALUE_REGEX)) {
                                currentPriceField.setText(oldValue);
                            }
                        });

        quantityField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.INVESTMENT_VALUE_REGEX)) {
                                quantityField.setText(oldValue);
                            }
                        });

        avgUnitPriceField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.INVESTMENT_VALUE_REGEX)) {
                                avgUnitPriceField.setText(oldValue);
                            }
                        });
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/base/BaseTickerManagement.kt
/*
 * Filename: BaseTickerManagement.kt (original filename: BaseTickerManagement.java)
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.base

import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.util.UIUtils
import org.moinex.model.enums.AssetType
import org.moinex.service.PreferencesService
import org.moinex.service.investment.TickerService

abstract class BaseTickerManagement(
    protected val tickerService: TickerService,
    protected val preferencesService: PreferencesService,
) {
    @FXML
    protected lateinit var nameField: TextField

    @FXML
    protected lateinit var symbolField: TextField

    @FXML
    protected lateinit var currentPriceField: TextField

    @FXML
    protected lateinit var quantityField: TextField

    @FXML
    protected lateinit var avgUnitPriceField: TextField

    @FXML
    protected lateinit var typeComboBox: ComboBox<AssetType>

    @FXML
    protected open fun initialize() {
        configureComboBoxes()
        populateTypeComboBox()
        configureListeners()
    }

    @FXML
    protected open fun handleCancel() {
        (nameField.scene.window as Stage).close()
    }

    @FXML
    protected abstract fun handleSave()

    protected fun populateTypeComboBox() {
        typeComboBox.items.setAll(*AssetType.entries.toTypedArray())
    }

    protected fun configureComboBoxes() {
        UIUtils.configureComboBox(typeComboBox, UIUtils::translateAssetType)
    }

    protected fun configureListeners() {
        currentPriceField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.INVESTMENT_VALUE_REGEX))) {
                currentPriceField.text = oldValue
            }
        }

        quantityField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.INVESTMENT_VALUE_REGEX))) {
                quantityField.text = oldValue
            }
        }

        avgUnitPriceField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.INVESTMENT_VALUE_REGEX))) {
                avgUnitPriceField.text = oldValue
            }
        }
    }
}
